vendor/src/github.com/docker/libnetwork/portmapper/mapper.go

```diff
@@ -128,8 +128,10 @@ func (pm *PortMapper) MapRange(container net.Addr, hostIP net.IP, hostPortStart,
 	}
 
 	containerIP, containerPort := getIPAndPort(m.container)
-	if err := pm.forward(iptables.Append, m.proto, hostIP, allocatedHostPort, containerIP.String(), containerPort); err != nil {
-		return nil, err
+	if hostIP.To4() != nil {
+		if err := pm.forward(iptables.Append, m.proto, hostIP, allocatedHostPort, containerIP.String(), containerPort); err != nil {
+			return nil, err
+		}
```

Comment:
> When starting docker with --userland-proxy=false, IPv6 connections don’t work at all: [..]

Yes that's the point I wanted to make. It will only work if the daemon was started with `--userland-proxy=false`. If we allow this to go in, it needs to be documented. Be aware there is a PR to have the proxy off by default. But as long as it is documented that user has to enable it for this feature, then we are good.

> I was under the impression that IPv6 requires the userland proxy.

That is not strictly true. If you have a routable IPv6 for your container, you can always route up to the container (Sure you may need to set some routes: [IPv6 with Docker](https://docs.docker.com/engine/userguide/networking/default_network/ipv6/) ). 

Also a general question, why do you need the port translation for IPv6 ?

> Using docker with IPv6 is completely different. One has to start docker daemon it with --fixed-cidr-v6= and access each container via it's own ipv6.

Not strictly true either. You can now create your bridge network witth

`docker network create --subnet <IPv6 pool> <network name>`